### PR TITLE
allow vsearch merge-pairs to output unmerged reads

### PR DESCRIPTION
**Pull request by colinvwood**

### CODE REVIEW

**Comment by gregcaporaso** [comment]
I'm realizing now that this is going to change the interface of `merge-pairs`. 

I see two ways to go here: 
1) Merge this just after the 2023.11 release, and include in the 2023.11 release notes that this interface will be changing in the first release of 2024.
2) We could add a parameter to `q2_vsearch._merge_pairs.merge_pairs` that toggles whether we create the unmerged reads output, and then have two actions that call that function - one that doesn't output the unmerged reads (i.e., the existing action) and a second action that does output the unmerged reads. 

I'm leaning toward Option 1, to avoid action-bloat. Thoughts on this @colinvwood, @ebolyen?

**Comment by colinvwood** [comment]
do we always warn of these sorts of changes one release cycle beforehand?

**Comment by gregcaporaso** [comment]
Generally, yes - we like to announce interface changes one cycle before making them (and this changes the interface by adding a new required output - so analysis workflows that call this action will fail until the new output is integrated in the workflow). That gives folks time to update any automated workflows.

**Comment by colinvwood** [comment]
I agree that the first option seems like the better way to go

**Comment by colinvwood** [comment]
I wonder though, since vsearch has never been part of the shotgun distribution, wouldn't adding it with these changes to that distribution not constitute a change of interface (but an introduction of one which is happening either way)? and the amplicon distribution can wait a cycle? 

but that might not be possible because we would need two branches on vsearch one for each distribution release

I guess we could release amplicon then merge this then release shotgun

**Comment by gregcaporaso** [comment]
Good point that it would be an interface change for the amplicon distro but not the shotgun distro. I'd be good with that plan, but I just wonder if it creates a lot of headaches for the release. Let's discuss in the engineering meeting today.

**Comment by colinvwood** [comment]
We're planning on merging just after the 2023.12 release for both distributions.

File: q2_vsearch/tests/test_merge_pairs.py [hunk 1]
```diff
@@ -87,11 +114,47 @@ def test_merge_pairs(self):
...
+                len(seq_lengths),
+                default_exp_sequence_counts[str(fastq_name)]
+            )
+
     def test_merge_pairs_some_samples_w_no_joined_seqs(self):
         # minmergelen is set very high here, resulting in only one sequence
         # being merged across the three samples.
         with redirected_stdio(stderr=os.devnull):
-            obs = merge_pairs(self.input_seqs, minmergelen=279)
+            obs, _ = merge_pairs(self.input_seqs, minmergelen=279)
```
Comment (gregcaporaso) [inline q2_vsearch/tests/test_merge_pairs.py:157]: This seems like a good place to check both the merged and unmerged files resulting from a single call to make sure that all input reads (by count) end up in one file or the other. I don't think you have that in one place elsewhere. I think that's a good test to include, whether it's in this specific test case or not.

File: q2_vsearch/plugin_setup.py [hunk 2]
```diff
@@ -282,7 +282,7 @@
         'minlen': qiime2.plugin.Int % qiime2.plugin.Range(0, None),
         'maxns': qiime2.plugin.Int % qiime2.plugin.Range(0, None),
         'allowmergestagger': qiime2.plugin.Bool,
-        'minovlen': qiime2.plugin.Int % qiime2.plugin.Range(0, None),
+        'minovlen': qiime2.plugin.Int % qiime2.plugin.Range(5, None),
```
Comment (gregcaporaso) [inline q2_vsearch/plugin_setup.py:285]: Why this change? I'm guessing vsearch enforces this so you're bumping it up?

File: q2_vsearch/tests/test_merge_pairs.py [hunk 3]
```diff
@@ -87,11 +114,47 @@ def test_merge_pairs(self):
...
+
+        # manifest is as expected
+        self._test_manifest_unmerged(unmerged)
+
+        # expected number of fastq files are created
+        output_fastqs = list(unmerged.sequences.iter_views(FastqGzFormat))
+        self.assertEqual(len(output_fastqs), 6)
+
+        default_exp_sequence_counts = {
+            'BAQ2687.1_0_L001_R1_001.fastq.gz': 31,
```
Comment (gregcaporaso) [inline q2_vsearch/tests/test_merge_pairs.py:129]: I'm guessing these numbers came from looking subtracted the expected merged sequence counts from the input sequence counts? Sounds good if so - just confirming.

**Review by gregcaporaso — CHANGES_REQUESTED** [review]
Looks great overall. Just a couple of questions regarding test data, and one request for a test.

File: q2_vsearch/plugin_setup.py (hunk 2 above)
Comment (colinvwood) [reply q2_vsearch/plugin_setup.py:285]: yep!

File: q2_vsearch/tests/test_merge_pairs.py (hunk 3 above)
Comment (colinvwood) [reply q2_vsearch/tests/test_merge_pairs.py:129]: They came from running vsearch on its own

File: q2_vsearch/tests/test_merge_pairs.py (hunk 1 above)
Comment (colinvwood) [reply q2_vsearch/tests/test_merge_pairs.py:157]: The `test_merge_pairs` and `test_unmerged_pairs` run the same command on the same data and assert the number of the merged and unmerged reads respectively. These numbers would sum to the number of records in the sample if one were to check. But this isn't necessary because the numbers come from vsearch directly.

**Review by gregcaporaso — APPROVED** [review]

**Comment by gregcaporaso** [comment]
Thanks @colinvwood!
